java/junit5/src/main/java/com/saucelabs/saucebindings/junit5/SauceBindingsExtension.java

```
@@ -1,114 +1,171 @@
-    this.dataCenter = dataCenter;
-  }
+  @Override
+  public void beforeEach(ExtensionContext context) {
+    SauceOptions options = updateOptions(context);
+    SauceSession session = new SauceSession(options);
+    session.setDataCenter(dataCenter);
+    WebDriver driver = session.start();
 
-  public SauceBindingsExtension(DataCenter dataCenter) {
-    this(new SauceOptions(), dataCenter);
+    getStore(context).put("session", session);
+    getStore(context).put("driver", driver);
   }
 
-  @Override
-  public void beforeEach(ExtensionContext context) {
-    if (isExtensionDisabled()) {
-      return;
+  private SauceOptions updateOptions(ExtensionContext context) {
+    SauceOptions options = sauceOptions.copy();
+    options.sauce().setName(context.getDisplayName());
+    options.sauce().getCustomData().put("sauce-bindings", "junit5");
+    List<String> tags = options.sauce().getTags();
+    if (tags != null) {
+      tags.addAll(context.getTags());
+    } else {
+      options.sauce().setTags(new ArrayList<>(context.getTags()));
     }
+    options.sauce().setBuild(buildName);
 
-    if (sauceOptions.sauce().getName() == null) {
-      sauceOptions.sauce().setName(context.getDisplayName());
-    }
+    return options;
+  }
 
-    session = new SauceSession(sauceOptions);
-    session.setDataCenter(dataCenter);
-    driver = session.start();
+  private ExtensionContext.Store getStore(ExtensionContext context) {
+    return context.getStore(
+        ExtensionContext.Namespace.create(getClass(), context.getRequiredTestMethod()));
   }
 
   @Override
   public void testSuccessful(ExtensionContext context) {
-    if (isExtensionDisabled()) {
-      return;
+    if (!SauceSession.isDisabled()) {
+      SauceSession session = (SauceSession) getStore(context).get("session");
+      RemoteWebDriver driver = session.getDriver();
+      try {
+        session.stop(true);
+      } catch (NoSuchSessionException e) {
+        LOGGER.severe(
```

Why `severe`? Is this breaking the test execution? Wouldn't a `warning` work better?

---

Because if you get this error it means the user code is preventing the library from being useful. If Sauce isn't getting pass/fail information, then they shouldn't be bothering with this library. So I don't see this as a caution but as an essential thing to be fixed.

---

We're going to need to document this if it's a change to how a user will perceive their results, or if it changes behavior they're used to. In general I agree, I just want to make sure we don't irritate people and make them switch back to non-bindings.

---

So 3 ways to do this:
1. Raise Exception - tell user this is absolutely not how to use the library
2. Log Severe - let the user do it but express quite strongly that it is really bad
3. Log Warning - tell users this isn't how they should be doing it

I can make an argument for why we should do 1, but I figured it would be better swallow the exception and complain loudly in the logs.